### PR TITLE
[TRTLLM-15757][refactor] Land the ModelRunner contract with the no-KVCache families

### PR DESCRIPTION
**Pull request by lori-ren**
## Dev Engineer Review

- Adds `ModelRunner`, `NoKVCacheRunner`, shared metadata helpers, and runner dispatch.
- Adds pooling and multimodal encoder execution paths.
- Updates `PyTorchModelEngine` delegation, warmup, forwarding, cleanup, CUDA-graph handling, and legacy fallback.
- Updates model comments to reference runner APIs.
- Review should verify runner selection, metadata parity, speculative decoding, CUDA graphs, multimodal validation, lifecycle behavior, and legacy paths.

## QA Engineer Review

- Adds tests for runner dispatch, metadata, input preparation, helper functions, pooling, multimodal encoding, forwarding, cleanup, and warmup.
- Updates CUDA-graph, prompt-lookahead, request-interface, and DSA warmup coverage.
- Removes obsolete multimodal, encode-only fallback, and duplicate CUDA-graph padding tests.
- No changed tests are listed in `tests/integration/test_lists/`, `test-db/`, or `qa/`.
- Coverage verdict: needs follow-up because the reported CI pipelines include failures and unstable results.

## Per-File QA Perspective

- `tensorrt_llm/_torch/pyexecutor/engine/metadata.py`: Adds attention and speculative-decoding metadata construction. Verify cache settings, draft limits, and dynamic-tree updates.
- `tensorrt_llm/_torch/pyexecutor/engine/runners/__init__.py`: Adds startup runner selection. Verify family resolution and unsupported-family fallback.
- `tensorrt_llm/_torch/pyexecutor/engine/runners/common.py`: Adds distributed, padding, multimodal, model-unwrapping, and position-offset helpers. Verify collective behavior and tensor transfers.
- `tensorrt_llm/_torch/pyexecutor/engine/runners/interface.py`: Adds shared runner contracts. Verify API consistency and dependency wiring.
- `tensorrt_llm/_torch/pyexecutor/engine/runners/no_kv_cache.py`: Adds shared no-KV-cache preparation and lifecycle handling. Verify input packing, metadata, LoRA, speculative decoding, and graph behavior.
- `tensorrt_llm/_torch/pyexecutor/engine/runners/pooling.py`: Adds pooling, classification, and reward-model forwarding. Verify logits handling and gathering.
- `tensorrt_llm/_torch/pyexecutor/engine/runners/mm_encoder.py`: Adds multimodal encoder forwarding. Verify request filtering, embedding lengths, output validation, and mRoPE propagation.
- `tensorrt_llm/_torch/pyexecutor/model_engine.py`: Changes engine startup, warmup, forwarding, cleanup, metadata, multimodal handling, and legacy fallback.
- `tensorrt_llm/_torch/models/modeling_*.py`: Updates internal runner API references in comments. No runtime behavior is expected.
- `tests/unittest/_torch/executor/engine/test_common.py`: Covers helper functions, distributed counts, padding, speculative metadata, offsets, and multimodal transfers.
- `tests/unittest/_torch/executor/engine/test_metadata.py`: Covers attention metadata and speculative-decoding updates.
- `tests/unittest/_torch/executor/engine/test_no_kv_cache.py`: Covers request packing, multimodal inputs, speculative decoding, and attention data parallelism.
- `tests/unittest/_torch/executor/engine/test_runners.py`: Covers dispatch, initialization, forwarding, lifecycle, pooling, and multimodal encoder behavior.
- `tests/unittest/_torch/executor/test_pytorch_model_engine.py`: Updates CUDA-graph, prompt-lookahead, cleanup, and DSA warmup tests, and removes obsolete fallback coverage.
- `tests/unittest/_torch/executor/test_pytorch_model_engine_warmup.py`: Covers runner lifecycle delegation and legacy warmup behavior.
- `tests/unittest/llmapi/test_llm_args.py`: Removes obsolete CUDA-graph padding tests.

### Motivation

`ModelEngine` owns execution details for every model family, making the coordinator difficult to extend. This PR introduces a runner boundary and starts moving family-specific no-KV-cache logic out of the engine.

### Protocol

Adds the `ModelRunner` protocol with `prepare_inputs`, `warmup`, `capture_graphs`, and `forward`. `PreparedInputs`, `RunnerConfig`, and `RunnerDeps` define the shared contracts.

### Migrated

Moves shared input preparation and metadata setup into `NoKVCacheRunner`. `PoolingRunner` and `MultimodalEncoderRunner` provide their family-specific forward steps and are selected at startup.

### Follow-up

Encoder-only, encoder-decoder, and decoder-only execution remain on their existing paths. Later migrations will remove the transitional exports, engine fallback, and injected model-forward callback.

## Test Coverage

- Shared no-KV-cache input preparation, metadata, helper, and lifecycle unit tests.
- `PoolingRunner` dispatch, engine delegation, model-forward, and logits-gathering unit tests.
- `test_e2e.py::test_ptp_quickstart_bert` leaves `encode_only` unset, so the standard `PyExecutor` dispatches BERT to `PoolingRunner` and compares its logits with Hugging Face.
- `accuracy/test_epd_disagg_multimodal.py::TestVideoMMEEPD::test_disaggregated_videomme` uses `mm_encoder_only` and validates EPD through `MultimodalEncoderRunner`.
- Full mypy check with compiled bindings.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why.
- PR follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work)).
- If PR introduces API changes, an appropriate PR label is added—either `api-compatible` or `api-breaking`.
- Any new dependencies have been scanned for license and vulnerabilities.
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) is updated if ownership changes.
- Documentation is updated as needed.
- The reviewers assigned automatically or manually are appropriate for this PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/engine/runners/mm_encoder.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                continue
+            mm_request_indices_with_payload.append(request_idx)
+            mm_params_with_payload.append(multimodal_param)
+            mm_embedding_lengths.append(multimodal_embedding_lengths)
+        if not mm_params_with_payload:
+            return {
+                "mm_embeddings": [],
+                "mm_embedding_request_indices": [],
+                "mm_embedding_lengths": [],
+            }
```
[the remark this answers is not in the document]
Comment (lori-ren) [reply tensorrt_llm/_torch/pyexecutor/engine/runners/mm_encoder.py:76]: Thanks for flagging this. The missing logits key is unchanged from the pre-refactor ModelEngine implementation. This PR is a behavior-preserving extraction, so it should be addressed in a follow-up PR rather than changing behavior here.

File: tensorrt_llm/_torch/pyexecutor/engine/runners/no_kv_cache.py [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+                    spec_metadata,
+                    [item[1] for item in all_rank_num_tokens],
+                    [item[2] for item in all_rank_num_tokens],
+                    [item[3] for item in all_rank_num_tokens],
+                )
+            else:
+                all_rank_num_tokens = self._deps.dist.tp_cp_allgather_int64(
+                    [attn_metadata.num_tokens]
+                )[:, 0].tolist()
+                attn_metadata.all_rank_num_tokens = all_rank_num_tokens
```
[the remark this answers is not in the document]
Comment (lori-ren) [reply tensorrt_llm/_torch/pyexecutor/engine/runners/no_kv_cache.py:367]: Thanks for flagging this. The Helix count overwrite and duplicate collective are unchanged from the pre-refactor _prepare_tp_inputs_no_cache implementation. This PR is a behavior-preserving extraction, so they should be addressed in a follow-up PR.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 3]
```diff
@@ -1069,6 +1059,54 @@ def __init__(
...
+            dist=self.dist,
+            mapping=self.mapping,
+            input_ids_cuda=self.input_ids_cuda,
+            position_ids_cuda=self.position_ids_cuda,
+            gather_ids_cuda=getattr(self, "gather_ids_cuda", None),
+            draft_tokens_cuda=getattr(self, "draft_tokens_cuda", None),
+            cache_indirection=self.cache_indirection_attention
+            if self.attn_backend.Metadata is TrtllmAttentionMetadata else None,
+            lora=self._lora,
+            model_forward=self.model_forward,
```
Comment (Shixiaowei02) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: Storing this bound method creates an engine-to-runner-to-engine reference cycle. PyExecutor.shutdown relies on del triggering immediate cleanup, so EPD teardown retains weights and GMS registrations until cyclic garbage collection, violating §6.5.

File: tests/unittest/_torch/executor/engine/test_runners.py [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+def test_resolve_runner_type_does_not_require_runtime_dependencies() -> None:
+    args = SimpleNamespace(encode_only=False, mm_encoder_only=False)
+
+    assert resolve_runner_type(_model(is_generation=False), args) is PoolingRunner
+
+
+def test_resolve_runner_checks_encode_only_before_pooling() -> None:
+    args = SimpleNamespace(encode_only=True, mm_encoder_only=False)
+
+    assert resolve_runner_type(_model(is_generation=False), args) is None
```
Comment (Shixiaowei02) [inline tests/unittest/_torch/executor/engine/test_runners.py:146]: The advertised BERT encoder-only end-to-end test remains on the legacy path here, never exercising either migrated runner. Please add EPD disaggregation coverage and document default-path validation under §6 before merging.

Comment (lori-ren) [reply tests/unittest/_torch/executor/engine/test_runners.py:146]: Thanks — the previous description conflated two BERT paths. The assertion here covers explicit `encode_only=True`, which intentionally bypasses `PyExecutor` and uses the existing `EncoderExecutor`, so no `ModelRunner` is selected. `test_ptp_quickstart_bert` does not set `encode_only`; it therefore builds the standard `PyExecutor`, and because BERT is non-generation, `resolve_runner_type` selects `PoolingRunner`. `PoolingRunner` is covered by dispatch, engine-delegation, model-forward, and logits-gathering unit tests, while the BERT test validates that complete default path against Hugging Face logits. The existing EPD VideoMME test sets `mm_encoder_only` through `MultimodalEncoder` and exercises `MultimodalEncoderRunner`. I updated the Test Coverage section to document both paths.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 3 above)
Comment (lori-ren) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Fixed by replacing the bound `model_forward` callback with an unbound method using a weak engine proxy, eliminating the reference cycle and preserving immediate teardown. This callback is a transitional workaround and will be removed once `DecoderRunner` and `EncoderRunner` are migrated.

File: tensorrt_llm/_torch/pyexecutor/engine/runners/no_kv_cache.py [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+class NoKVCacheRunner(ABC):
+    """Common execution template for model families that do not use a KV cache."""
+
+    def __init__(
+        self,
+        model: nn.Module,
+        deps: RunnerDeps,
+        config: NoKVCacheRunnerConfig,
+    ) -> None:
+        self._model = model
```
Comment (QiJune) [inline tensorrt_llm/_torch/pyexecutor/engine/runners/no_kv_cache.py:67]: [P1] Release the runner-held model before release_gc()

The weak proxy fixes the runner-to-engine reference cycle, but the runner still keeps the model alive through self._model. During terminal teardown, PyTorchModelEngine.__del__() calls cleanup() while self._runner is still alive, so setting self.model = None does not free the weights before release_gc() calls torch.cuda.empty_cache(). The runner and model are released only after __del__ returns; in a long-lived worker this can leave a weights-sized CUDA allocator reservation and OOM the next model load. Please clear or close self._runner before clearing self.model and calling release_gc(), and add a lifecycle test verifying that the model is unreachable when release_gc() runs.

**Review by QiJune — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/pyexecutor/engine/runners/no_kv_cache.py (hunk 5 above)
Comment (lori-ren) [reply tensorrt_llm/_torch/pyexecutor/engine/runners/no_kv_cache.py:67]: Fixed in 8d2193febf: `cleanup()` now clears `_runner` before `release_gc()`, with a lifecycle test verifying the runner-held model is released first.
